src/main/java/com/google/sps/data/UserDatastoreRepository.java

```diff
@@ -0,0 +1,130 @@
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.gson.Gson;
+import com.google.sps.data.User;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+
+/**
+ * This class implements the match repository using the datastore 
+ */
+public class UserDatastoreRepository implements UserRepository {
+    
+  private Collection<User> allUsers = new ArrayList<>();
+  
+  private static final Gson gson = new Gson();
+
+  private DatastoreService datastore;
+  
+  private int maxProfiles = 10;
+  
+  public UserDatastoreRepository() {
+      datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  // function to add profile to database
+  public void addUserToDatabase(User user) {
+    String name = user.toString();
+    Collection<String> specialties = user.getSpecialties();
+
+    Entity userEntity = new Entity("User");
+    userEntity.setProperty("name", name);
+    if(specialties.size() > 0){
+        userEntity.setProperty("specialties", specialties);
+    }
+    datastore.put(userEntity);
+  }
+
+  // function to fetch the user from the database
+  public Collection<User> fetchUserProfiles() {
+    Query query = new Query("User");
+    PreparedQuery results = datastore.prepare(query);
+    List<Entity> resultsList = results.asList(FetchOptions.Builder.withLimit(maxProfiles));
+
+    Collection<User> userProfiles = new ArrayList<>();
+    for (Entity entity : resultsList) {
+        String name = (String) entity.getProperty("name");
+        Collection<String> specialties = (Collection) entity.getProperty("specialties");
+
+        User userObject = new User(name);
+        if(specialties != null) {
+            for (String specialty : specialties) {
+                userObject.addProductArea(specialty);
+            }
+        }
+        userProfiles.add(userObject);
+    }
+    return userProfiles;
+  }
+
+  public void removeUserProfile(User user){
+    Query query = new Query("User");
+    PreparedQuery results = datastore.prepare(query);
+    List<Entity> resultsList = results.asList(FetchOptions.Builder.withLimit(maxProfiles));
+
+    String name = user.toString();
+
+    for (Entity entity : resultsList) {
+        String currentName = (String) entity.getProperty("name");
+        if(name.contains(currentName)) {
+            Key current = entity.getKey();
+            datastore.delete(current);
+        }
+    }
+  }
+
+  public void addUser(User user) {
```

Comment:
Maybe add a function comment documenting that nothing happens if the user already exists?

Reply:
What did you mean by this?

Reply:
Do you mean add a comment above the function explaining it and describing what happens is user exists?

Reply:
Ok I did that

Reply:
You got it! A user of addUser might expect it to overwrite what was there, so it's kind to inform that this isn't the case (that the new data is just dropped).